### PR TITLE
relax enforcement of options sets in validation

### PR DESCRIPTION
**Pull request by replay**
this relaxes the input validation because we have seen issues where queries using deprecated aggregation functions have been rejected (f.e. `sumSeries`). We want these deprecated functions to still be accepted, while logging a warning that a deprecated function has been used.

#2508

### CODE REVIEW

File: webapp/graphite/functions/params.py [hunk 1]
```diff
@@ -135,11 +136,12 @@ def validateParams(func, params, args, kwargs):
...
 
-    # parameter is restricted to a defined set of values, but value is not in it
+    # parameter is restricted to a defined set of possible values, but given value is not in it.
+    # possibly it's a deprecated but still accepted value.
     if params[i].options and value not in params[i].options:
-      raise InputParameterError(
-        'Invalid option specified for function "{func}" parameter "{param}"'.format(
-          param=params[i].name, func=func))
+      log.warning(
+        'given option "{option}" in function "{func}" is not in list of valid options, it may be deprecated but still accepted'
```
Comment (DanCech) [inline webapp/graphite/functions/params.py]: I'd still include the parameter name here, and shorten up the text:
```
'Deprecated value "{value}" specified for parameter "{param}" of function "{func}"'
```
Longer-term we pshould support stricter enforcement for cases where there is a well-understood set of valid options, but this should be fine for now.  Most likely we'd need to add another flag to the param to set the enforcement level, and/or a separate set of deprecated values that are accepted but trigger the log message.

Comment (replay) [reply webapp/graphite/functions/params.py]: Ok, but at that point we can't be certain if the given value is indeed a deprecated value or if it's a completely invalid value. So if the message says it's deprecated, that might also lead to confusing situations or not? How about:

```
Invalid or deprecated value "{value}" specified for parameter "{param}" of function "{func}"'
```

**Comment by Dieterbe** [comment]
this doesn't make sense to me. the whole reason to introduce the validation is to change http 5xx responses into 4xx ones. Then we made validation optional. Now we're changing it such that even when validation is enabled, we don't do validation, but just log a warning, and we treat every case the same? why aren't we precisely checking whether the provided value is valid, deprecated, or invalid?

also, if invalid values make it through, do we return 5xx again?

**Comment by replay** [comment]
this was discussed here https://github.com/graphite-project/graphite-web/issues/2508

**Comment by replay** [comment]
I would like to move forward with this in some way. This PR is implementing what's described in https://github.com/graphite-project/graphite-web/issues/2508#issuecomment-547147780

The validation is not bullet proof and there will probably always be some cases that slip through. One thing which we could do to lower the likely hood of 5xx due to invalid options is to only disable the validation of options if the parameter name is `func`. That would be kind of weirdly inconsistent though, and it might be confusing to users who define a custom graphite function with a parameter named `func`, so i'd prefer to stay consistent.

**Comment by DanCech** [comment]
`aggFunc` is already a dedicated parameter type, so the simplest solution will be to implement handling for that in the validator so it can allow deprecated options to be used.  The cleanest and most extensible way to do that would be to allow each `ParamType` to define its own validation function that can then be called from the central validator.

**Comment by replay** [comment]
@DanCech what do you think about this solution? I didn't make a dedicated `ParamType` for each of the possible parameter types, but I registered the parameter types via a class method `register` which takes a name and an optional validator and instantiates the type.
Since the case of the `aggFuncs` is a bit special, because it's getting the list of valid values from `graphite.render.functions`, it has its own sub-class of `ParamType` called `ParamTypeAggFunc` which can be registered in the same way as the other `ParamType` instances. 

The only thing that i kind of dislike about this solution is the fact that from `graphite.render.functions` we need to call `ParamTypeAggFunc.setValidAggFuncs()` to set the valid function names.

**Comment by DanCech** [comment]
We're going to move the definition of the `safe*` functions, `aggFuncs`, `aggAliases` and `getAggFunc` out of `functions.py` and into the `functions` directory so we can import them directly in `params.py` without creating a circular import.

File: webapp/graphite/functions/params.py [hunk 2]
```diff
@@ -2,29 +2,31 @@
...
-  'nodeOrTag',
-  'series',
-  'seriesList',
-  'seriesLists',
-  'string',
-  'tag',
-]:
-  setattr(ParamTypes, paramType, paramType)
+class ParamType(object):
+  name = 'undefined'
```
Comment (DanCech) [inline webapp/graphite/functions/params.py]: we don't need this afaik

Comment (replay) [reply webapp/graphite/functions/params.py]: True, it will always be defined by the constructor anyway

**Pull request merged**